Importing Necessary Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Reading the dataset

In [3]:
df=pd.read_csv(r"D:\UPDATED_NLP_COURSE\UPDATED_NLP_COURSE\05-Topic-Modeling\npr.csv")

In [4]:
df.head()

Article
0  In the Washington of 2016, even when the polic...
1    Donald Trump has used Twitter  —   his prefe...
2    Donald Trump is unabashedly praising Russian...
3  Updated at 2:50 p. m. ET, Russian President Vl...
4  From photography, illustration and video, to d...

Shape of Dataset

In [5]:
df.shape

(11992, 1)

It has 11992 rows

Let' check for null values and blanks present in dataset

In [7]:
df.isnull().sum()

Article    0
dtype: int64

There are no null values present in the data

In [11]:
blanks=[]

for i in df['Article']:
    if i.isspace():
        blanks.append(i)

In [12]:
blanks

[]

There are no blanks rows present in the data

## Now our data is good to go with implementing LATENT DIRICHLET ALLOCATION to define topics of articles

## Pre-Processing

We have to use Count Vectorizer for LDA, Importing Count Vectorizer and LDA

In [14]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

We are initialising max_df as 0.9, which indicates, it ignores the words that appears in the documents more than 90%.

and min_df=2, indicates that, it doesn't consider the words that appears in the documents less than 2 documents

In [15]:
countv=CountVectorizer(max_df=0.9,min_df=2,stop_words='english')

In [25]:
cv=countv.fit_transform(df['Article'])

In [24]:
len(countv.vocabulary_)

54777

The number of words in vocabulary is 54777

### Now, We are implementing the LDA technique. We are taking topics of 7 to lable the Articles

In [26]:
lda=LatentDirichletAllocation(n_components=7,random_state=42,verbose=2)

In [27]:
lda.fit(cv)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   51.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 1 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   46.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 2 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   41.7s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 3 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   37.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 4 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   34.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 5 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   31.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 6 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   29.7s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 7 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   29.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 8 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   28.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 9 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   27.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


iteration: 10 of max_iter: 10


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   25.3s finished


LatentDirichletAllocation(n_components=7, random_state=42, verbose=2)

In [33]:
len(lda.components_)

7

We have got the fitted data for 7 components.



In [36]:
lda.components_[0].shape

(54777,)

Topic 0 has 54777 vales , which is in equal to length of vocabulary

In [39]:
lda.components_[0].argsort()[-10:]

array([33390, 36310, 21228, 10425, 31464,  8149, 36283, 22673, 42561,
       42993], dtype=int64)

## Now, we take each component(topic) data and find the top 20 words.

In [54]:
for i, j in enumerate(lda.components_):
    print(f"Top 20 Words that appeared in each Topic:{i}")
    print([countv.get_feature_names()[k] for k in lda.components_[i].argsort()[-20:]])
    print('\n')

Top 20 Words that appeared in each Topic:0
['president', 'state', 'tax', 'insurance', 'trump', 'companies', 'money', 'year', 'federal', '000', 'new', 'percent', 'government', 'company', 'million', 'care', 'people', 'health', 'said', 'says']


Top 20 Words that appeared in each Topic:1
['white', 'according', 'attack', 'reported', 'war', 'military', 'house', 'security', 'russia', 'government', 'npr', 'reports', 'says', 'news', 'people', 'told', 'police', 'president', 'trump', 'said']


Top 20 Words that appeared in each Topic:2
['little', 'know', 'don', 'year', 'make', 'way', 'world', 'family', 'home', 'day', 'time', 'water', 'city', 'new', 'years', 'food', 'just', 'people', 'like', 'says']


Top 20 Words that appeared in each Topic:3
['world', 'research', 'university', 'percent', 'care', 'time', 'new', 'don', 'years', 'medical', 'disease', 'patients', 'just', 'children', 'study', 'like', 'women', 'health', 'people', 'says']


Top 20 Words that appeared in each Topic:4
['donald', 'politi

Above 20 words are more occured in the each topic.



## Now, we will create a new column and assign lable numbers based on max value

In [56]:
cv.shape

(11992, 54777)

From the above shape output, we can clearly observer that, 11992 articles which are treated as rows and 54777 columns, which is vocabulary length after doing count vectorisation.

Now, we are tranforming the data by applying with LDA.

In [58]:
lab=lda.transform(cv)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   25.3s finished


In [62]:
lab.shape

(11992, 7)

The shape of our data is 11992 rows and 7 columns

In [67]:

df['Label']=lab.argmax(axis=1)

In [68]:
df.head()

Article  Label
0  In the Washington of 2016, even when the polic...      1
1    Donald Trump has used Twitter  —   his prefe...      1
2    Donald Trump is unabashedly praising Russian...      1
3  Updated at 2:50 p. m. ET, Russian President Vl...      1
4  From photography, illustration and video, to d...      2

## Now, We map the labels with label category that article belongs to

In [69]:
mapper={0:'Politics',1:'Securtiy',2:'General',3:'Health',4:'Election',5:'Music',6:'Education'}

In [70]:
df['Lable_Name']=df['Label'].map(mapper)

In [71]:
df.head()

Article  Label Lable_Name
0  In the Washington of 2016, even when the polic...      1   Securtiy
1    Donald Trump has used Twitter  —   his prefe...      1   Securtiy
2    Donald Trump is unabashedly praising Russian...      1   Securtiy
3  Updated at 2:50 p. m. ET, Russian President Vl...      1   Securtiy
4  From photography, illustration and video, to d...      2    General

This work of defining Topics is completely inferred, not exactly known like supervised ones.

This comes under unsupervised learning.

Initially, we considered 7 components, which are topics.This components are basically initiated with domain knowledge.

#### Thank You